skorch/callbacks/training.py

```diff
@@ -273,3 +278,155 @@ def _sink(self, text, verbose):
+    history is not saved along with the model. To avoid this, in case
+    you use ``ParamMapper`` (or subclasses, e.g. :class:`Initializer`)
+    and want to save your model make sure to either (a) use pickle,
+    (b) save and load the history or (c) remove the parameter mapper
+    callbacks before continuing training.
+
+    Examples
+    --------
+    Initialize a layer on first epoch before the first training step:
+
+    >>> init = partial(torch.nn.init.uniform_, a=0, b=1)
+    >>> cb = ParamMapper('linear*.weight', at=1, fn=init)
+    >>> net = Net(myModule, callbacks=[cb])
+
+    Reset layer initialization if train loss reaches a certain value:
+
+    >>> at = lambda net: net.history[-1, 'train_loss'] > 1
+    >>> init = partial(torch.nn.init.uniform_, a=0, b=1)
+    >>> cb = ParamMapper('linear0.weight', at=at, fn=init)
+    >>> net = Net(myModule, callbacks=[cb])
+
+    Periodically freeze and unfreeze all embedding layers:
+
+    >>> def my_sched(net, _at, _fn):
+    ...    if len(net.history) % 2 == 0:
+    ...        return skorch.utils.freeze_parameter
+    ...    else:
+    ...        return skorch.utils.unfreeze_parameter
+    >>> cb = ParamMapper('embedding*.weight', schedule=my_sched)
+    >>> net = Net(myModule, callbacks=[cb])
+
+    Parameters
+    ----------
+    patterns : str or callable or list
+      The pattern(s) to match parameter names against. Patterns are
+      UNIX globbing patterns as understood by :func:`~fnmatch.fnmatch`.
+      Patterns can also be callables which will get called with the
+      parameter name and are regarded as a match when the callable
+      returns a truthy value.
+
+      Example: ``'linear*.weight'`` or ``['linear0.*', 'linear1.bias']``
+            or ``lambda name: name.startswith('linear')``.
+
+    fn : function
+      The function to apply to each parameter separately.
+
+    at : int or callable
+      In case you specify an integer it represents the epoch number the
+      function ``fn`` is applied to the parameters, in case ``at`` is
+      a function it will receive ``net`` as parameter and the function
+      is applied to the parameter once ``at`` returns ``True``.
+
+    schedule : callable or None
+      If specified it determines how ``fn`` and ``at`` are used.
+      This function receives ``net``, ``at`` and ``fn`` and returns
+      the function to be applied on the matched parameters.
+
+    """
+    def __init__(self, patterns, fn=noop, at=1, schedule=None):
+        self.at = at
+        self.fn = fn
+        self.schedule = schedule or self._default_schedule
+        self.patterns = patterns
+
+    def initialize(self):
+        if not isinstance(self.patterns, (list, tuple)):
+            self.patterns = [self.patterns]
+
+        if isinstance(self.at, int):
+            self.at = partial(self._epoch_at, epoch=self.at)
+
+    def named_parameters(self, net):
+        return list(net.module_.named_parameters())
```

Comment:
The following should work without needing to convert the generator to a list:
```python
def named_parameters(self, net):
    yield from net.module_.named_parameters()
```

Reply:
If `named_parameters` is only consumed once (which seems to be the case), a simple `return net.module_.parameters()` should be enough (or no special method at all).

Reply:
Pretty sure that it is necessary since we evaluate it more than once (for each pattern)

Reply:
Ah I see.

Reply:
I was thinking something like this for `filter_parameters`:

```python
from collections import product

def filter_parameters(self, patterns, params):
    patten_fns = [
        pattern if callable(pattern) else lambda n: fnmatch(n, pattern)
    ]
    for pattern_fn, (name, param) in product(patten_fns, params):
        if pattern_fn(name):
            yield name, param
```

This would allow `named_parameters` to remain a generator.